### PR TITLE
Handle command completions in RTE

### PR DESCRIPTION
**Pull request by alunturner**
This PR adds slash command functionality to the rich text editor. It does this by:

- allowing suggestions from the rust model that start with a "/" to build a query which is passed to the `<Autocomplete />` component
- selecting a command from the list puts that text inside the rte and closes the modal
- when the message is sent, there is a pile of extra logic in the `sendMessage` function that then handles a slash command if present

Example of use (previously starting a menu with "/" did not open any autocomplete):

https://user-images.githubusercontent.com/56027671/230392473-f4493ea4-ef39-4de8-9128-ef9dfe40a1ee.mov

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Handle command completions in RTE ([\#10521](https://github.com/matrix-org/matrix-react-sdk/pull/10521)). Contributed by @alunturner.

### CODE REVIEW

File: src/components/views/rooms/wysiwyg_composer/hooks/useEditing.ts [hunk 1]
```diff
@@ -45,11 +45,12 @@ export function useEditing(
...
-        () =>
-            !!mxClient && content !== undefined && editMessage(content, { roomContext, mxClient, editorStateTransfer }),
-        [content, roomContext, mxClient, editorStateTransfer],
-    );
+    const editMessageMemoized = useCallback(async () => {
+        if (mxClient === undefined || content === undefined) {
+            return;
+        }
+        return editMessage(content, { roomContext, mxClient, editorStateTransfer });
+    }, [content, roomContext, mxClient, editorStateTransfer]);
```
Comment (alunturner) [inline src/components/views/rooms/wysiwyg_composer/hooks/useEditing.ts:53]: This was previously returning `false | Promise<...>` which I don't think is what we were aiming for here.

File: src/components/views/rooms/wysiwyg_composer/hooks/useWysiwygSendActionHandler.ts [hunk 2]
```diff
@@ -46,6 +46,7 @@ export function useWysiwygSendActionHandler(
 
             switch (payload.action) {
                 case "reply_to_event":
+                case Action.FocusAComposer:
```
Comment (alunturner) [inline src/components/views/rooms/wysiwyg_composer/hooks/useWysiwygSendActionHandler.ts:49]: This is the action that is fired when a user enters a bad command and then makes a decision from the dialog they're presented with - it ensures that we put the cursor back in the composer when the dialog closes.

File: src/components/views/rooms/wysiwyg_composer/utils/message.ts [hunk 3]
```diff
@@ -71,9 +76,51 @@ export async function sendMessage(
     }*/
     PosthogAnalytics.instance.trackEvent<ComposerEvent>(posthogEvent);
 
-    const content = await createMessageContent(message, isHTML, params);
+    let content: IContent | null = null;
 
-    // TODO slash comment
+    // Functionality here approximates what can be found in SendMessageComposer.sendMessage()
```
Comment (alunturner) [inline src/components/views/rooms/wysiwyg_composer/utils/message.ts:81]: The logic in this block is not really new, it has been taken from the current composer's code that is run when a message is sent (can be seen in `SendMessageComposer.tsx`), with changes made where appropriate to pass correct variables to function calls etc

File: src/components/views/rooms/wysiwyg_composer/hooks/useWysiwygSendActionHandler.ts (hunk 2 above)
Comment (robintown) [reply src/components/views/rooms/wysiwyg_composer/hooks/useWysiwygSendActionHandler.ts:49]: Is that documented in the actions file? Would be good to if not

File: src/components/views/rooms/wysiwyg_composer/utils/message.ts [hunk 4]
```diff
@@ -71,9 +76,51 @@ export async function sendMessage(
...
+    if (message.startsWith("/") && !message.startsWith("//")) {
+        const { cmd, args } = getCommand(message);
+        if (cmd) {
+            // we will need to handle /me separately, see SlashCommands.tsx:1387
```
Comment (robintown) [inline src/components/views/rooms/wysiwyg_composer/utils/message.ts:85]: It seems a little brittle to refer to a line number like this, is it possible to just describe the location?

**Review by robintown — APPROVED** [review]
Looks good otherwise :+1:

File: src/components/views/rooms/wysiwyg_composer/utils/message.ts (hunk 4 above)
Comment (alunturner) [reply src/components/views/rooms/wysiwyg_composer/utils/message.ts]: Yep the action is already documented as it's what is used by the current composer. Fair point on the comment, I couldn't figure out a way to make it a useful link so I've just described the location more generally.

**Pull request merged**